File: insolation.c
#include "ebm.h"

void calc_ins(const SimulationConfig *conf, const PlanetParams *para,
              ClimateState *state, double ins[]) {
  int reso = conf->reso;
  int n;
  double Q = para->Q;
  double delta, h, cosH, H, r_AU, day;
  double theta[reso];

  for (n = 0; n < reso; n++) {
    theta[n] = (n - 90.0) * M_PI / 180.0;
  }

  // Calculate day based on season
  day = (state->season - fmod(state->season, conf->day_sec)) / conf->day_sec;

  // Calculate hour angle h
  h = fmod(state->season, conf->day_sec) * 2.0 * M_PI / conf->day_sec - M_PI;

  calc_delta(conf, para, state->season, &delta, &r_AU);

  for (n = 0; n < reso; n++) {
    cosH = -tan(theta[n]) * tan(delta);
    if (cosH >= 1.0)
      H = 0.0;
    else if (cosH <= -1.0)
      H = M_PI;
    else
      H = acos(cosH);

    if (r_AU != 0.0) {
      ins[n] = (Q / M_PI / r_AU / r_AU) * (H * sin(theta[n]) * sin(delta) +
                                           cos(theta[n]) * cos(delta) * sin(H));
    } else {
      fprintf(stderr, "r_AU is negative!");
    }
  }
}

void calc_Yearsec(double *Ysec) {
  double ma = 227936640000, G = 6.67384e-11, M = 1.9891e30, m = 639e21;
  *Ysec = sqrt(ma * ma * ma / (G * (M + m))) * 2.0 * M_PI;
}

void calc_delta(const SimulationConfig *conf, const PlanetParams *para,
                double season, double *delta, double *r_AU) {
  int loop = 0, loop_max = 100;
  double u, r;
  double ma = 227936640000.0, G = 6.67384e-11, M = 1.9891e30, m = 639e21;
  double ecc = 0.0934, p = 336.049 * M_PI / 180.0, oneAU = 149597871000.0;
  double x, nt, cosf, sinf;

  nt = sqrt(G * (M + m) / ma / ma / ma) * season;
  if (ecc == 0.0)
    u = nt;
  else {
    x = nt;
    do {
      x = x - (x - ecc * sin(x) - nt) / (1.0 - ecc * cos(x));
      loop = loop + 1;
      // REVERTED to match legacy behavior (removed fabs)
    } while (x - ecc * sin(x) - nt >= 1.0e-6 && loop < loop_max);
    u = x;
  }

  r = ma * (1.0 - ecc * cos(u));
  *r_AU = r / oneAU;

  if (ecc != 0.0) {
    cosf = (ma * (1.0 - ecc * ecc) / r - 1.0) / ecc;
    sinf = sqrt(1.0 - cosf * cosf);
    if (sin(u) < 0.0)
      sinf = -1.0 * sinf;
  } else if (ecc == 0.0) {
    cosf = cos(u);
    sinf = sin(u);
  }
  *delta = asin(sin(para->obl) * (sinf * cos(p) + cosf * sin(p)));
}

void calc_ins_s(const SimulationConfig *conf, const PlanetParams *para,
                ClimateState *state, double alpha, double ins_s[]) {
  int reso = conf->reso;
  int n, mark = 0;
  double Q = para->Q;
  double delta, h, cosH_t, H_t, cosH_s, H_s, r_AU, H_eff;
  double theta[reso], slope[reso];

  calc_delta(conf, para, state->season, &delta, &r_AU);
  h = fmod(state->season, conf->day_sec) * 2.0 * M_PI / conf->day_sec - M_PI;

  for (n = 0; n < reso; n++) {
    theta[n] = (n - 90.0) * M_PI / 180.0;
    slope[n] = theta[n] + alpha;

    cosH_t = -tan(theta[n]) * tan(delta);
    if (cosH_t >= 1.0)
      H_t = 0.0;
    else if (cosH_t <= -1.0)
      H_t = M_PI;
    else
      H_t = acos(cosH_t);

    if (slope[n] > 0.5 * M_PI) {
      slope[n] = M_PI - slope[n];
      mark = 1;
    }
    if (slope[n] < -0.5 * M_PI) {
      slope[n] = -M_PI - slope[n];
      mark = 2;
    }
    cosH_s = -tan(slope[n]) * tan(delta);
    if (cosH_s >= 1.0)
      H_s = 0.0;
    else if (cosH_s <= -1.0)
      H_s = M_PI;
    else
      H_s = acos(cosH_s);
    if (mark == 1) {
      slope[n] = M_PI - slope[n];
      mark = 0;
    }
    if (mark == 2) {
      slope[n] = -M_PI - slope[n];
      mark = 0;
    }

    H_eff = fmax(H_t, H_s); // REVERTED to match legacy behavior (fmax)

    if (h >= -H_eff && h <= H_eff) {
      // REVERTED to match legacy behavior (Q / r_AU / r_AU without M_PI)
      ins_s[n] =
          Q / r_AU / r_AU *
          (sin(slope[n]) * sin(delta) + cos(slope[n]) * cos(delta) * cos(h));
      if (ins_s[n] <= 0.0)
        ins_s[n] = 0.0;
    } else {
      ins_s[n] = 0.0;
    }
  }
}

File: haskell-migration/ebm-mars/src/EBM/Insolation.hs
{-|
Module      : EBM.Insolation
Description : Insolation (solar radiation) calculations
Copyright   : (c) 2026
License     : MIT

Functions for calculating solar insolation on Mars, including
orbital mechanics and slope-dependent radiation.
-}

module EBM.Insolation
  ( calcYearLength
  , calcDelta
  , calcInsolation
  , calcSlopeInsolation
  ) where

import EBM.Types
import qualified Data.Vector.Unboxed as V

-- Physical constants
semiMajorAxis :: Double
semiMajorAxis = 227936640000.0  -- [m]

gravitationalConstant :: Double
gravitationalConstant = 6.67384e-11  -- [m^3 kg^-1 s^-2]

sunMass :: Double
sunMass = 1.9891e30  -- [kg]

marsMass :: Double
marsMass = 639e21  -- [kg]

eccentricity :: Double
eccentricity = 0.0934

perihelion :: Double
perihelion = 336.049 * pi / 180.0  -- [rad]

oneAU :: Double
oneAU = 149597871000.0  -- [m]

-- | Calculate Martian year length [s]
calcYearLength :: Time
calcYearLength =
  let ma = semiMajorAxis
      g = gravitationalConstant
      m = sunMass
      mm = marsMass
  in Time $ sqrt (ma * ma * ma / (g * (m + mm))) * 2.0 * pi

-- | Calculate orbital parameters (declination and distance)
-- Returns (delta [rad], r_AU)
calcDelta :: SimulationConfig -> PlanetParams -> Time -> (Angle, Double)
calcDelta config params season =
  let ma = semiMajorAxis
      g = gravitationalConstant
      m = sunMass
      mm = marsMass
      ecc = eccentricity
      p = perihelion
      
      nt = sqrt (g * (m + mm) / ma / ma / ma) * unTime season
      
      -- Solve Kepler's equation iteratively
      u = if ecc == 0.0
          then nt
          else let x0 = nt
                   x1 = x0 - (x0 - ecc * sin x0 - nt) / (1.0 - ecc * cos x0)
               in solveKepler ecc nt x1 1 100
      
      r = ma * (1.0 - ecc * cos u)
      rAU = r / oneAU
      
      -- Calculate true anomaly
      (cosf, sinf) = if ecc /= 0.0
                     then let cf = (ma * (1.0 - ecc * ecc) / r - 1.0) / ecc
                              sf = sqrt (1.0 - cf * cf)
                          in if sin u < 0.0
                             then (cf, -sf)
                             else (cf, sf)
                     else (cos u, sin u)
      
      -- Calculate declination
      delta = asin (sin (unAngle $ obliquity params) * (sinf * cos p + cosf * sin p))
      
  in (Angle delta, rAU)
  where
    -- Solve Kepler's equation: x - ecc * sin(x) - nt = 0
    -- CRITICAL: Match C version exactly (no fabs in condition)
    solveKepler :: Double -> Double -> Double -> Int -> Int -> Double
    solveKepler ecc nt x loop loopMax
      | loop >= loopMax = x
      | x - ecc * sin x - nt >= 1.0e-6 = 
          let xNew = x - (x - ecc * sin x - nt) / (1.0 - ecc * cos x)
          in solveKepler ecc nt xNew (loop + 1) loopMax
      | otherwise = x

-- | Floating-point modulo (C fmod equivalent)
fmod :: Double -> Double -> Double
fmod x y = x - fromIntegral (floor (x / y)) * y

-- | Calculate global insolation
calcInsolation :: SimulationConfig -> PlanetParams -> ClimateState -> V.Vector Double
calcInsolation config params state =
  let reso = resolution config
      q = solarConstant params
      theta = V.generate reso (\n -> (fromIntegral n - 90.0) * pi / 180.0)
      
      -- Calculate hour angle h (matching C version exactly)
      h = (unTime (season state) `fmod` unTime (dayLength config)) * 2.0 * pi / unTime (dayLength config) - pi
      
      (delta, rAU) = calcDelta config params (season state)
      
  in V.generate reso $ \n ->
       let th = theta V.! n
           cosH = -tan th * tan (unAngle delta)
           hAngle = if cosH >= 1.0
                    then 0.0
                    else if cosH <= -1.0
                    then pi
                    else acos cosH
       in if rAU /= 0.0
          then (q / pi / rAU / rAU) * 
               (hAngle * sin th * sin (unAngle delta) + 
                cos th * cos (unAngle delta) * sin hAngle)
          else 0.0

-- | Calculate slope insolation
calcSlopeInsolation :: SimulationConfig -> PlanetParams -> ClimateState -> Angle -> V.Vector Double
calcSlopeInsolation config params state alpha =
  let reso = resolution config
      q = solarConstant params
      
      (delta, rAU) = calcDelta config params (season state)
      h = (unTime (season state) `fmod` unTime (dayLength config)) * 2.0 * pi / unTime (dayLength config) - pi
      
      theta = V.generate reso (\n -> (fromIntegral n - 90.0) * pi / 180.0)
      
  in V.generate reso $ \n ->
       let th = theta V.! n
           slope0 = th + unAngle alpha
           
           cosHt = -tan th * tan (unAngle delta)
           hT = if cosHt >= 1.0
                then 0.0
                else if cosHt <= -1.0
                then pi
                else acos cosHt
           
           -- Adjust slope angle if out of range
           (slope, mark) = if slope0 > 0.5 * pi
                           then (pi - slope0, 1)
                           else if slope0 < -0.5 * pi
                           then (-pi - slope0, 2)
                           else (slope0, 0)
           
           cosHs = -tan slope * tan (unAngle delta)
           hS = if cosHs >= 1.0
                then 0.0
                else if cosHs <= -1.0
                then pi
                else acos cosHs
           
           -- Restore slope if it was adjusted
           slopeFinal = case mark of
                          1 -> pi - slope
                          2 -> -pi - slope
                          _ -> slope
           
           hEff = max hT hS  -- C version uses fmax
           
       in if h >= -hEff && h <= hEff
          then let ins = q / rAU / rAU *  -- C version: no division by pi
                         (sin slopeFinal * sin (unAngle delta) + 
                          cos slopeFinal * cos (unAngle delta) * cos h)
               in if ins <= 0.0 then 0.0 else ins
          else 0.0
